src/data/finset.lean

```diff
@@ -659,6 +659,29 @@ begin
   { intro x, simp, intros hx hx₂, refine ⟨or.resolve_left (h hx) hx₂, hx₂⟩ }
 end
 
+/- We can simplify an application of filter where the decidability is inferred in "the wrong way" -/
+@[simp] lemma filter_congr_decidable {α} (s : finset α) (p : α → Prop) (h : decidable_pred p)
+  [decidable_pred p] : @filter α p h s = s.filter p :=
```

Comment:
Does any mechanism in `simp` prevents it from applying this lemma without making progress?

Reply:
Good question!

The answer seems to be yes: `simp` will check whether the application of a simp-lemma made progress. If I run
```lean
example {α} (s : finset α) (p q : α → Prop) [decidable_pred p] [decidable_pred q] :
  s.filter p = s.filter q :=
by squeeze_simp
```
then `filter_congr_decidable` is not listed in the output (and I do if I change LHS to `{x ∈ s | p x }`).

Reply:
Nice!